### PR TITLE
feat: Mouse picking

### PR DESCRIPTION
**Pull request by PaablooCH**
Now the mouse picking uses the quadtree

### CODE REVIEW

**Review by DADSILVER — APPROVED** [review]

**Comment by gerard-queralt** [comment]
It appears you added/updated some assets and binaries. Are those changes intended?

**Comment by markmeliaa** [comment]
> It appears you added/updated some assets and binaries. Are those changes intended?

Yes, it was because we modifed some meshes in order to check if the raycast was behaving correctly, do we need to delete those?

**Comment by gerard-queralt** [comment]
> Yes, it was because we modifed some meshes in order to check if the raycast was behaving correctly, do we need to delete those?

I don't think there's any need, just wanted to make sure it was intended

File: Source/DataStructures/Quadtree.cpp [hunk 1]
```diff
@@ -506,4 +508,47 @@ std::list<const GameObject*> Quadtree::GetAllGameObjects(const GameObject* gameO
 		familyObjects.insert(familyObjects.end(), objectsChildren.begin(), objectsChildren.end());
 	}
 	return familyObjects;
+}
+
+void Quadtree::CheckRaycastIntersection(std::map<float, const GameObject*>& hitGameObjects, const LineSegment& ray)
+{
+	std::queue<const Quadtree*> quadtreeQueue;
+	quadtreeQueue.push(this);
+
```
Comment (OriolSors) [inline Source/DataStructures/Quadtree.cpp:517]: Perfect non-recursive implementation!

File: Source/Modules/ModuleRender.cpp [hunk 2]
```diff
@@ -203,6 +201,8 @@ update_status ModuleRender::Update()
 		skybox->Draw();
 	}
 
+	gameObjectsToDraw.clear();
+
```
Comment (OriolSors) [inline Source/Modules/ModuleRender.cpp:205]: Do we need the cleaning in the Update instead of PreUpdate?

**Review by OriolSors — APPROVED** [review]
As we discussed during the last weeks, all seems to be agreed in the short term. I find more than good the approach of using the Quadtree to speed up the raycasting but taking into account the current selected object too. It is a functional and efficient solution but (maybe) not the cleanest one... We need to clarify and stablish all together what will the cases and limits of the Quadtree be, and then polish this (only if needed). But for now, perfect job all of you!!!

PD: If you can, answer my question commented. I don't know if this change was made during our discussion about using the RenderList instead of Quadtree GO, or if it is needed for some reason I miss right now. Thanks in advance!!

File: Source/DataStructures/Quadtree.cpp (hunk 1 above)
Comment (markmeliaa) [reply Source/DataStructures/Quadtree.cpp:517]: Thanks!

File: Source/Modules/ModuleRender.cpp (hunk 2 above)
Comment (markmeliaa) [reply Source/Modules/ModuleRender.cpp:205]: Oh, it was my bad, fixed it!

Comment (OriolSors) [reply Source/Modules/ModuleRender.cpp:205]: Fine!! It was because of trying to raycast to the RenderList right?

Comment (markmeliaa) [reply Source/Modules/ModuleRender.cpp]: Yes! I was about to answer your other comment, it was left there when we tried to work with the RenderList, but forgot to changed it back later.

Comment (OriolSors) [reply Source/Modules/ModuleRender.cpp]: Perfect!!

**Comment by markmeliaa** [comment]
> As we discussed during the last weeks, all seems to be agreed in the short term. I find more than good the approach of using the Quadtree to speed up the raycasting but taking into account the current selected object too. It is a functional and efficient solution but (maybe) not the cleanest one... We need to clarify and stablish all together what will the cases and limits of the Quadtree be, and then polish this (only if needed). But for now, perfect job all of you!!!

I agree that maybe a better solution could be possible, but yea, for now, this will do the work :)

**Comment by markmeliaa** [comment]
Ready to be merged! @gerard-queralt 🫡

**Pull request merged**